mm/kfence.c

```diff
@@ -448,48 +459,57 @@ static void kfence_print_stack(int obj_index, bool is_alloc)
 	}
 }
 
-static void kfence_dump_object(int obj_index)
+static void kfence_dump_object(int obj_index, struct alloc_metadata *obj)
 {
-	int size = abs(kfence_metadata[obj_index].size);
-	unsigned long start = kfence_index_to_addr(obj_index);
+	int size = abs(obj->size);
+	unsigned long start = kfence_obj_to_addr(obj, obj_index);
 
 	pr_err("Object #%d: starts at %px, size=%d\n", obj_index, (void *)start,
 	       size);
 	pr_err("allocated at:\n");
-	kfence_print_stack(obj_index, true);
+	kfence_print_stack(obj, true);
 	if (kfence_metadata[obj_index].state == KFENCE_OBJECT_FREED) {
 		pr_err("freed at:\n");
-		kfence_print_stack(obj_index, false);
+		kfence_print_stack(obj, false);
 	}
 }
 
-static inline void kfence_report_oob(unsigned long address, int obj_index)
+static inline void kfence_report_oob(unsigned long address, int obj_index,
+				     struct alloc_metadata *object)
 {
-	unsigned long object = kfence_index_to_addr(obj_index);
-	bool is_left = address < object;
+	bool is_left = address < kfence_obj_to_addr(object, obj_index);
 
 	pr_err("==================================================================\n");
 	pr_err("BUG: KFENCE: slab-out-of-bounds at address %px to the %s of object #%d\n",
 	       (void *)address, is_left ? "left" : "right", obj_index);
 	dump_stack();
-	kfence_dump_object(obj_index);
+	kfence_dump_object(obj_index, object);
 	pr_err("==================================================================\n");
 }
 
-static inline void kfence_report_uaf(unsigned long address, int obj_index)
+static inline void kfence_report_uaf(unsigned long address, int obj_index,
+				     struct alloc_metadata *object)
 {
 	pr_err("==================================================================\n");
 	pr_err("BUG: KFENCE: use-after-free at address %px on object #%d\n",
 	       (void *)address, obj_index);
 	dump_stack();
-	kfence_dump_object(obj_index);
+	kfence_dump_object(obj_index, object);
 	pr_err("==================================================================\n");
 }
 
+static inline void copy_obj_metadata(int index, struct alloc_metadata *obj)
+{
+	if ((index < 0) || (index >= KFENCE_NUM_OBJ))
+		memset(obj, 0, sizeof(struct alloc_metadata));
+	memcpy(obj, &kfence_metadata[index], sizeof(struct alloc_metadata));
```

Comment:
*obj = kfence_metadata[index]
should be simpler and safer.

Reply:
Ack.